### PR TITLE
Simplifies the state stack system.

### PR DESCRIPTION
**Pull request by Hanamuke**
Change the State info implementation to a memory contiguous stack,
member of Position. To save a few more ops, gamePly is replace with
only the ply of the base (basePly) that need not be updated during search.

StatePtrList is no needed anymore.
StateInfo is now a private member structure of Position, 
It is completely abstracted away from the rest of the code

http://tests.stockfishchess.org/tests/view/5ab5474f0ebc590295d88476

LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 98519 W: 20889 L: 20916 D: 56714

Non functional.

### CODE REVIEW

**Comment by Hanamuke** [comment]
I don't have strong feelings about this patch actually, do you like it or not ?

**Comment by mcostalba** [comment]
Apart from extremely slowing down position's copy, I don't see any other reason to move an array of more than 200 StateInfo objects inside Position class.

**Comment by mcostalba** [comment]
Sorry, not 'to move' but 'to create' because currently there is no such array and memory management is done in fully automatic and naturally extends to multi threads, without gimmicks

**Comment by mcostalba** [comment]
A slightly better approach (but not as good as current master) would be to move the preallocated array under Thread class.

**Comment by snicolet** [comment]
Indeed the stack is a property of the search, and not of a chess position, let's not mix the two.

**Comment by Hanamuke** [comment]
@mcostalba Genuine question, does thread setting up time (position copy mostly) consume allocated search time ? I could make copy fast. The overhead would only be rule50+1 StateInfo copies.
I do not understand you criticism about "gimmicks" though. All I did for multithreading was to delete 
16 lines from threads.cpp

@snicolet  What if i followed the suggestion of marco to move stateStack inside Thread ? I completely agree that it would be better.

The Position copy overhead could be the reason why my test struggled to pass even though i measured 0.5% speedup on the bench. It was at the extremely short TC and multithreading. If some machines were running DDR3 it would explain why the red and yellow workers were mostly only workers with more losses.

I could close the PR and do a new one after polishing the patch if you would consider merging it with those changes. I believe that powerful abstraction is the mark of good software design and i don't think that's it's normal that all functions that use do_move hae to create local StateInfo, when with the exception of pliesFromNull and pieceCaptured it only consists of strictly Position related information. Besides it is unsafe/bad practice that the StateInfo* st; member of Position points to temporary objects created by callers of do_move();

Finally, for the sake of being partial, i'd like to point out two drawbacks of master made worse by the patch : 
-Issue 1 : uci_loop has a local Position objects that is useless during search and takes up memory. I could probably be solved by making uci_loop use the rootPos of the main thread but i haven't looked into it.

-Issue 2 : in tbprobe and in endgame, Position objects are created solely to get material keys. It was already a waste of time and memory compared with a dedicated function, but now it uses even more memory.

These issues obvisouly weren't a concern since noone voiced them out but they could be dealt with in a future PR.

What do you want me to do @snicolet ?

**Comment by mcostalba** [comment]
@Hanamuke you copy positions every time you start a search, with 8 threads you copy 8 positions for every move. Regarding gimmicks, I was referring to [this](https://github.com/official-stockfish/Stockfish/pull/1508/files#diff-760a56b9dcfc15bf67577d7d6a273fb9R982).

**Comment by mcostalba** [comment]
@Hanamuke pre-alocating an array under Threads is not a simplification, is a design choice, hence is good as long as is better/faster than current solution (also in SMP case).

**Comment by mcostalba** [comment]
@Hanamuke @snicolet talking about design choices: preallocating an array it means to fix maximum game length (not only search depth). For instance allocating [100 + MAX_PLY] it means that if a game is longer than 100 moves than it _could_ crash the engine if search goes all the way to near MAX_PLY.

To me this would be a really poor design choice compared to current (that supports unlimited game length).

**Comment by Hanamuke** [comment]
@mcostalba You haven't understood setup_moves(). It's not for multithreading, but precisely to avoid stack overflowing for long games. 102 is a value chosen because of the 50 moves rule. No matter how long the game is, setup_moves only registers the setup states after the last zeroing move, hence maximum 100 plies. I need one more because i make the move before checking if it zeroes and one more because the search can reach MAX_PLY, making a MAX_PLY +1 states to remember. 101 would probably be ok since the last setup pos and the root pos are the same.

**Comment by syzygy1** [comment]
@snicolet 
The StateInfo stack is not the search stack.

(However, I personally have no objection to merging StateInfo, search stack and MovePicker objects and to moving search-related stuff that for some reason is part of the Thread class (why?) to the Position class (which maybe should then be called Tree class or something, but it's just a name) :-))

I'm not expressing an opinion on this particular PR, because I haven't fully studied it.

**Comment by syzygy1** [comment]
If we are into substantial design changes, it could also be considered to allocate one large MoveList for each Position object. That should be more cache friendly than the current approach. It doesn't necessarily give simpler code, but perhaps most of the changes can be hidden.

**Comment by Hanamuke** [comment]
@syzygy1 thank you for the idea, i'm indeed more into low level/algorithm kind of stuff, i work in operations research.

File: src/position.cpp [hunk 1]
```diff
@@ -159,12 +158,21 @@ void Position::init() {
   Zobrist::noPawns = rng.rand<Key>();
 }
 
+/// Position::set() copies a Position object and sets it to a thread objet
```
Comment (vondele) [inline src/position.cpp]: object

File: src/position.cpp [hunk 2]
```diff
@@ -966,10 +973,22 @@ void Position::undo_null_move() {
 
   assert(!checkers());
 
-  st = st->previous;
+  --st;
   sideToMove = ~sideToMove;
 }
 
+/// Position::setup_move() rebase the state stack when a move
```
Comment (vondele) [inline src/position.cpp]: unclear comment

File: src/position.h [hunk 3]
```diff
@@ -184,7 +153,35 @@ class Position {
   int castlingRightsMask[SQUARE_NB];
   Square castlingRookSquare[CASTLING_RIGHT_NB];
   Bitboard castlingPath[CASTLING_RIGHT_NB];
-  int gamePly;
+
+/// StateInfo struct stores information needed to restore a Position object to
+/// its previous state when we retract a move. Whenever a move is made on the
+/// board (by calling Position::do_move), a StateInfo object must be passed.
```
Comment (vondele) [inline src/position.h]: ```a StateInfo object must be passed``` still valid ?

File: src/position.h [hunk 4]
```diff
@@ -184,7 +153,35 @@ class Position {
...
+
+  // Not copied when making a move (will be recomputed anyhow)
+  Key        key;
+  Bitboard   checkersBB;
+  Piece      capturedPiece;
+  Bitboard   blockersForKing[COLOR_NB];
+  Bitboard   pinners[COLOR_NB];
+  Bitboard   checkSquares[PIECE_TYPE_NB];
+};
+  /// A stack that keep track of the state history of the position.
```
Comment (vondele) [inline src/position.h]: ```102``` needs a comment

**Comment by vondele** [comment]
overall I like the patch, even if some things can be further clarified. The fact that one can make a move on a position without the need of passing something else, or query a position for draw etc, makes sense. Having said that, it contains some non-trivial changes, which needs some more study, maybe some more elegant approach is possible?

As @syzygy1 has pointed out, this is not the search state, it is not mixing things IMO, but providing the needed information to describe a position completely (including draw by repetition etc.)... 

Slowdown seems not an issue, as it passed [-3,1].

**Comment by mcostalba** [comment]
@Hanamuke I hate to repeat myself, but I see that maybe we have a communication issue here.

I have fully understand what setup_moves() stands for. Thank you.

Regarding copying position, of course current master copies too when start a search...but it does not copy all the new StateInfo array. Moving this under Thread class would avoid copying the array at every move. Moreover it is also that place that makes most sense, given that it is a per-thread object, populated during search (like history, etc).

**Comment by Hanamuke** [comment]
I will move the stack over to Thread tonight. Since you think setup time is crucial, i will try to blend in some optimization by adding setup information to ThreadPool.

**Comment by Hanamuke** [comment]
Turns out migrating the stack to Thread, or reintroducing the "StateInfo* previous" pointer to avoid copies, is too complicated/unnatural, this is not the right way to do it. 

The current version make Position safer by not binding st to a temporary object, and let the user use do_move without creating StateInfo objects.  It also simplifies thread creation. This is at the cost of some (imo negligeable) copy overhead at thread start, and the use of setup_moves. 

As some pointed out, the state stack is different from the search stack, and it mainly contains strictly positional information. 

@snicolet If this is not acceptable, you can close the PR.

**Comment by mcostalba** [comment]
Let's write some number:

```
sizeof(StateInfo)  = 176
sizeof(stateStack) = 40128

```

So with, say 8 threads game, we **gratuitously** copy 120K of new additional data at every move in the game!

Why we have to do this? For some synctactic sugar? This is a heavy pessimization of the algoritm used to do/undo moves withouth any benefit. It is worst than current by performance and by design (the latter one is subjective but to me, moving history of states into Position class, adding tricks as setup_moves() just to avoid overflow and using a fixed global array instead of leveraging the natural recursion in search() it means to move from a good design to a lower design).

**Comment by Hanamuke** [comment]
@mcostalba This is a gross exaggeration, i only copy the setup states which is at most 100 StateInfo per thread, but on average rule50 is very low so 20 StateInfo  / 3.5Ko per thread is a better average estimate. 

Besides, i don't even know what you're arguing for since it's simpler and any possible slowdown was shown to be negligible since it passed non regression.

**Comment by mcostalba** [comment]
@Hanamuke please let me express my view of how this PR comment stuff work.

- Everybody is free to open a PR.

- Everybody is free to comment on them as long as he has something **on topic** to say.

- Maintainer will decide what to do.

I don't comment here because I was previous maintainer, I comment here as a simple contributor, as anybody can do and does.

So please:

- Please don't take it personally, it is not.
- Please accept that everybody can comment.
- Invoking the maintainer last decision does not stop people commenting (also because it does not work in this way)

**Comment by Hanamuke** [comment]
I did a final optimization to position copy, i only copy one StateInfo per thread, and only the keys for other setup positions, so it will be only ~300o copied per thread.

**Comment by snicolet** [comment]
@Hanamuke Is the patch ready for review? If so, I will print the code and study it calmly, as it is quite a complex pull request.

**Comment by Hanamuke** [comment]
@snicolet Yeah it is ready, for review. Sorry for the late response my mail box is buggy atm (laposte...). I'll just do a last commit to merge with master and correct an assert, but nothing that would affect the review.

**Comment by Hanamuke** [comment]
Btw stéphane, i don't think anyone has tested the PR with TBs. The changes in tbprobe.cpp should be non functional but if you want to be on the safe side...
I'm available if you have questions.

**Pull request closed without merging**